File: src/MyServiceBus.Testing/TestingServiceExtensions.cs
using System;
using Microsoft.Extensions.DependencyInjection;

namespace MyServiceBus;

public static class TestingServiceExtensions
{
    public static IServiceCollection AddServiceBusTestHarness(this IServiceCollection services, Action<IBusRegistrationConfigurator> configure)
    {
        var configurator = new BusRegistrationConfigurator(services);
        configure(configurator);

        configurator.Build();

        services.AddSingleton<InMemoryTestHarness>();
        services.AddSingleton(typeof(IConsumerFactory<>), typeof(ScopeConsumerFactory<>));
        services.AddSingleton<IMessageBus>((sp) => sp.GetRequiredService<InMemoryTestHarness>());
        services.AddSingleton<ITransportFactory>((sp) => sp.GetRequiredService<InMemoryTestHarness>());
        services.AddSingleton<IReceiveEndpointConnector>((sp) => (IReceiveEndpointConnector)sp.GetRequiredService<IMessageBus>());
        services.AddScoped(typeof(IRequestClient<>), typeof(GenericRequestClient<>));
        services.AddScoped<IRequestClientFactory, RequestClientFactory>();

        return services;
    }
}

File: src/Java/myservicebus-testing/src/main/java/com/myservicebus/TestingServiceExtensions.java
package com.myservicebus;

import com.myservicebus.di.ServiceCollection;
import java.util.function.Consumer;

public class TestingServiceExtensions {
    public static ServiceCollection addServiceBusTestHarness(ServiceCollection services,
            Consumer<BusRegistrationConfigurator> configure) {
        var configurator = new BusRegistrationConfiguratorImpl(services);
        configure.accept(configurator);
        configurator.complete();

        services.addSingleton(InMemoryTestHarness.class, sp -> () -> new InMemoryTestHarness(sp));
        services.addSingleton(TransportSendEndpointProvider.class,
                sp -> () -> (TransportSendEndpointProvider) sp.getService(InMemoryTestHarness.class));
        services.addSingleton(RequestClientTransport.class,
                sp -> () -> (RequestClientTransport) sp.getService(InMemoryTestHarness.class));
        services.addScoped(ScopedClientFactory.class,
                sp -> () -> new RequestClientFactory(sp.getService(RequestClientTransport.class)));

        return services;
    }
}

